cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go

```diff
@@ -0,0 +1,595 @@
+)
+
+// TerraformProvider exposes infrastructure provisioning using Azure Terraform templates
+type TerraformProvider struct {
+	env         *environment.Environment
+	projectPath string
+	options     Options
+	console     input.Console
+	cli         terraform.TerraformCli
+	azCli       azcli.AzCli
+}
+
+type TerraformDeploymentDetails struct {
+	ParameterFilePath  string
+	PlanFilePath       string
+	localStateFilePath string
+}
+
+// Name gets the name of the infra provider
+func (t *TerraformProvider) Name() string {
+	return "Terraform"
+}
+
+func (t *TerraformProvider) RequiredExternalTools() []tools.ExternalTool {
+	return []tools.ExternalTool{t.cli}
+}
+
+// NewTerraformProvider creates a new instance of a Terraform Infra provider
+func NewTerraformProvider(ctx context.Context, env *environment.Environment, projectPath string, infraOptions Options) *TerraformProvider {
+	terraformCli := terraform.GetTerraformCli(ctx)
+	console := input.GetConsole(ctx)
+	azCli := azcli.GetAzCli(ctx)
+
+	// Default to a module named "main" if not specified.
+	if strings.TrimSpace(infraOptions.Module) == "" {
+		infraOptions.Module = "main"
+	}
+
+	return &TerraformProvider{
+		env:         env,
+		projectPath: projectPath,
+		options:     infraOptions,
+		console:     console,
+		cli:         terraformCli,
+		azCli:       azCli,
+	}
+}
+
+// Previews the infrastructure through terraform plan
+func (t *TerraformProvider) Plan(ctx context.Context) *async.InteractiveTaskWithProgress[*DeploymentPlan, *DeploymentPlanningProgress] {
+	return async.RunInteractiveTaskWithProgress(
+		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeploymentPlan, *DeploymentPlanningProgress]) {
+			os.Setenv("TF_DATA_DIR", t.dataDirPath())
+			isRemoteBackendConfig, err := t.isRemoteBackendConfig()
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("reading backend config: %w", err))
+				return
+			}
+
+			modulePath := t.modulePath()
+
+			t.console.Message(ctx, "Initializing terraform...")
+			err = asyncContext.Interact(func() error {
+				initRes, err := t.init(ctx, isRemoteBackendConfig)
+				if err != nil {
+					return fmt.Errorf("terraform init failed: %s , err: %w", initRes, err)
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				asyncContext.SetError(err)
+				return
+			}
+
+			t.console.Message(ctx, "\nGenerating terraform parameters...")
+			currentSubscription, err := t.ensureEnvSubscription(ctx)
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("failed to set az subscription , err: %w", err))
+				return
+			}
+			if currentSubscription != "" {
+				defer t.setAZSubscription(ctx, currentSubscription)
+			}
+
+			err = CreateInputParametersFile(t.parametersTemplateFilePath(), t.parametersFilePath(), t.env.Values)
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("creating parameters file: %w", err))
+				return
+			}
+
+			t.console.Message(ctx, "Validating terraform template...")
```

Comment:
In the `console.Message()` above, we prepended the string with \n.  In this message, we don't put in a line break at all, and in the next message we put it at the end.   I was surprised to see the difference, is that intentional?

Reply:
Basically to play nice with the terraform `stdout` output already written so these messages don't just merge in and get lost with terraforms wall of text.

Reply:
Ahh - that makes sense.